### PR TITLE
[core] Extract raylet registration into core worker

### PR DESCRIPTION
**Pull request by dentiny**
Necessary for issue: https://github.com/ray-project/ray/issues/47994

Since we will combine (1) worker registration and (2) port announcement, before that we need to extract (1) into core worker.

Update: the code is expected to be a no-op refactor, so the code structure is at a cleaner state before we merge APIs.

### CODE REVIEW

**Comment by dentiny** [comment]
Hi @jjyao can I get a review for this PR? It's been a week, thanks!

File: src/ray/core_worker/core_worker.h [hunk 1]
```diff
@@ -1355,6 +1355,19 @@ class CoreWorker : public rpc::CoreWorkerServiceHandler {
   FRIEND_TEST(TestOverrideRuntimeEnv, TestCondaInherit);
   FRIEND_TEST(TestOverrideRuntimeEnv, TestCondaOverride);
 
+  /// Register core worker to worker pool.
+  Status RegisterWorkerToWorkerPool(raylet::RayletConnection &conn,
```
[the remark this answers is not in the document]
Comment (dentiny) [reply src/ray/core_worker/core_worker.h]: Updated.

File: src/ray/raylet_client/raylet_client.h [hunk 2]
```diff
@@ -256,45 +256,11 @@ class RayletClientInterface : public PinObjectsInterface,
...
-  void ShutdownIfLocalRayletDisconnected(const Status &status);
-  /// The connection to raylet.
-  std::shared_ptr<ServerConnection> conn_;
-  /// A mutex to protect stateful operations of the raylet client.
-  std::mutex mutex_;
-  /// A mutex to protect write operations of the raylet client.
-  std::mutex write_mutex_;
-};
-
 class RayletClient : public RayletClientInterface {
```
Comment (rynewang) [inline src/ray/raylet_client/raylet_client.h:259]: can we have a comment here about what's difference of RayletClient vs RayletConnection ? Something like: RayletConnection is a simple connection, whereas RayletClient sends Register (and Announce) RPCs in ctor to mark this core worker as ready (IIUC)

File: src/ray/core_worker/core_worker.cc [hunk 3]
```diff
@@ -362,33 +406,33 @@ CoreWorker::CoreWorker(CoreWorkerOptions options, const WorkerID &worker_id)
...
-                                             &local_raylet_id,
-                                             &assigned_port,
-                                             options_.serialized_job_config,
-                                             options_.startup_token);
-
+  auto raylet_conn = std::make_unique<raylet::RayletConnection>(
+      io_service_, options_.raylet_socket, /*num_retries=*/-1, /*timeout=*/-1);
+  NodeID local_raylet_id;
+  int assigned_port = 0;
+  Status raylet_client_status = RegisterWorkerToRaylet(*raylet_conn,
```
Comment (rynewang) [inline src/ray/core_worker/core_worker.cc:413]: well, looks like RayletClient is not for sending RPC either, it's done here by CoreWorker. Then why do we have 2 classes RayletClient and RayletConnection? or, why do we want a *public* RayletConnection?

Comment (dentiny) [reply src/ray/core_worker/core_worker.cc:413]: I didn't change any accessibility for `RayletConnection` class.

Comment (rynewang) [reply src/ray/core_worker/core_worker.cc:413]: right it's not about visibility but intended API usage. Now, the caller (core worker) have to create a RayletConnection then use it to create a RayletClient, without any real usage of RayletConnection itself. Maybe we can cave a `static RayletClient Create()` func?

File: src/ray/raylet_client/raylet_client.h (hunk 2 above)
Comment (dentiny) [reply src/ray/raylet_client/raylet_client.h:261]: I added doc for raylet connection.

Comment (dentiny) [reply src/ray/raylet_client/raylet_client.h:261]: > RayletClient sends Register (and Announce) RPCs in ctor to mark this core worker as ready (IIUC)

raylet client is way more complicated than that, checkout the interface
https://github.com/ray-project/ray/blob/1b07eaf9163853bfa429f86328443db30aee903b/src/ray/raylet_client/raylet_client.h#L222-L227

Tried to add a simple wording.

File: src/ray/core_worker/core_worker.cc (hunk 3 above)
Comment (dentiny) [reply src/ray/core_worker/core_worker.cc:413]: > Now, the caller (core worker) have to create a RayletConnection then use it to create a RayletClient, without any real usage of RayletConnection itself.

I'm not sure what you mean, it's used in two places:
- `RegisterWorkerToRaylet` takes a raylet connection reference
- Then ownership passed to raylet client and used inside

**Comment by dentiny** [comment]
Hi @rynewang wondering if you have further feedbacks?

**Review by rynewang — APPROVED** [review]

**Pull request merged**